src/v/kafka/security/acl.h

```diff
@@ -0,0 +1,474 @@
+ */
+enum class principal_type {
+    user,
+};
+
+inline std::ostream& operator<<(std::ostream& os, resource_type type) {
+    switch (type) {
+    case resource_type::topic:
+        return os << "topic";
+    case resource_type::group:
+        return os << "group";
+    case resource_type::cluster:
+        return os << "cluster";
+    case resource_type::transactional_id:
+        return os << "transactional_id";
+    }
+}
+
+inline std::ostream& operator<<(std::ostream& os, pattern_type type) {
+    switch (type) {
+    case pattern_type::literal:
+        return os << "literal";
+    case pattern_type::prefixed:
+        return os << "prefixed";
+    }
+}
+
+/*
+ * Kafka principal is (principal-type, principal)
+ */
+class acl_principal {
+public:
+    acl_principal(principal_type type, ss::sstring name)
+      : _type(type)
+      , _name(std::move(name)) {}
+
+    friend bool operator==(const acl_principal&, const acl_principal&)
+      = default;
+
+    template<typename H>
+    friend H AbslHashValue(H h, const acl_principal& e) {
+        return H::combine(std::move(h), e._type, e._name);
+    }
+
+    bool wildcard() const { return _name == "*"; }
+
+private:
+    principal_type _type;
+    ss::sstring _name;
+};
+
+inline const acl_principal acl_wildcard_user(principal_type::user, "*");
+
+/*
+ * Resource pattern matches resources using a (type, name, pattern) tuple. The
+ * pattern type changes how matching occurs (e.g. literal, name prefix).
+ */
+class resource_pattern {
+public:
+    static constexpr const char* wildcard = "*";
+
+    resource_pattern(resource_type type, ss::sstring name, pattern_type pattern)
+      : _resource(type)
+      , _name(std::move(name))
+      , _pattern(pattern) {}
+
+    friend bool operator==(const resource_pattern&, const resource_pattern&)
+      = default;
+
+    template<typename H>
+    friend H AbslHashValue(H h, const resource_pattern& e) {
+        return H::combine(std::move(h), e._resource, e._name, e._pattern);
+    }
+
+    friend std::ostream& operator<<(std::ostream&, const resource_pattern&);
+
+    resource_type resource() const { return _resource; }
+    const ss::sstring& name() const { return _name; }
+    pattern_type pattern() const { return _pattern; }
+
+private:
+    resource_type _resource;
+    ss::sstring _name;
+    pattern_type _pattern;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const resource_pattern& r) {
+    fmt::print(
+      os,
+      "type {{{}}} name {{{}}} pattern {{{}}}",
+      r._resource,
+      r._name,
+      r._pattern);
+    return os;
+}
+
+/*
+ * A host (or wildcard) in an ACL rule.
+ */
+class acl_host {
+public:
+    explicit acl_host(const ss::sstring& host)
+      : _addr(host) {}
+
+    explicit acl_host(ss::net::inet_address host)
+      : _addr(host) {}
+
+    static acl_host wildcard_host() { return acl_host{}; }
+
+    bool wildcard() const { return _addr.has_value(); }
```

Comment:
This looks inverted

Reply:
oof good catch

Reply:
oh, it's unused. that's why it didn't effect anything